bazaarvoice/emodb
# Use cell tombstones in blocked delta tables

## Github Issue #

`None`

## What Are We Doing Here?

In our QA environments, we have observed tombstone issues in blocked tables. Unlike the old delta tables, the new delta tables use range tombstones for deletes instead of cell tombstones. Because of a bug in Cassandra 2.1, we cannot use tracing to observe how many tombstones exist in a given row. 

My best guess as to why the blocked tables are having tombstone issues is that range tombstones are less efficient for our use than cell tombstones. This PR alters our compaction process to use cell tombstones whenever a delta has 2 or fewer blocks.

This PR continues to use range tombstones on the blocked tables whenever the `migrationStage` is set read from the old tables. This is because it extremely complicated to compute the number blocks in a delta if we did not read it from the blocked table. We should first test this PR in environments that are already reading from the blocked tables and see if it improves performance. If we are successful, I can then make a different pull request with adds support for cell tombstones when reading from the legacy delta tables.

## How to Test and Verify

1. Check out this PR
2. Set `migrationPhase` to `DOUBLE_WRITE_BLOCKED_READ`
3. Write large deltas to Emo and ensure they get compacted properly under the new model
4. Additionally, set `DOUBLE_WRITE_LEGACY_READ` and ensure that the old tables are unharmed by this change

## Risk

### Level 

`High`

### Required Testing

The whole kitchen sink!

### Risk Summary

This is pretty high risk, as we are modifying the way Emo does compaction on the new table. If something is wrong with it, we could potentially cause a corruption.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
